modules/shared_data/src/main/scala/org/elpaca_metagraph/shared_data/daemons/fetcher/YouTubeFetcher.scala

```diff
@@ -0,0 +1,149 @@
+package org.elpaca_metagraph.shared_data.daemons.fetcher
+
+import cats.effect.Async
+import cats.syntax.all._
+import fs2.io.net.Network
+import org.elpaca_metagraph.shared_data.app.ApplicationConfig
+import org.elpaca_metagraph.shared_data.calculated_state.CalculatedStateService
+import org.elpaca_metagraph.shared_data.types.DataUpdates.YouTubeUpdate
+import org.elpaca_metagraph.shared_data.types.Refined.ApiUrl
+import org.elpaca_metagraph.shared_data.types.States.{DataSourceType, YouTubeDataSource}
+import org.elpaca_metagraph.shared_data.types.YouTube.LatticeClient.{LatticeUser, LatticeUsersApiResponse}
+import org.elpaca_metagraph.shared_data.types.YouTube.YouTubeDataAPI.{SearchListResponse, VideoDetails, VideoListResponse}
+import org.http4s.circe.jsonOf
+import org.http4s.client.Client
+import org.http4s.{Method, Request, Uri}
+import org.tessellation.node.shared.resources.MkHttpClient
+
+import java.time.format.DateTimeFormatter
+import java.time.{Duration, LocalDateTime, ZoneId, ZoneOffset}
+
+class YouTubeFetcher[F[_] : Async : Network](apiKey: String, baseUrl: ApiUrl)(implicit client: Client[F]) {
+  def fetchDagUsers(apiUrl: ApiUrl, offset: Long = 0, users: List[LatticeUser] = List.empty): F[List[LatticeUser]] =
+    client.expect[LatticeUsersApiResponse](Request[F](Method.GET, Uri.unsafeFromString(apiUrl.toString())
+      .withQueryParam("limit", 100)
+      .withQueryParam("offset", offset)))(jsonOf[F, LatticeUsersApiResponse]).flatMap { response =>
+      val newUsers = users ++ response.data
+      if (!response.meta.exists(meta => meta.offset + meta.limit < meta.total)) {
+        newUsers.filter(user => user.primaryDagAddress.isDefined && user.youtube.isDefined).pure
+      } else fetchDagUsers(apiUrl, response.meta.get.offset + response.meta.get.limit, newUsers)
+    }
+
+  def fetchVideosByChannelId(channelId: String,
+                             searchString: String,
+                             minimumDuration: Long,
+                             minimumViews: Long,
+                             publishedAfter: Option[LocalDateTime] = None,
+                             pageToken: Option[String] = None,
+                             result: List[VideoDetails] = List.empty): F[List[VideoDetails]] = {
+    val formattedPublishAfter = publishedAfter.map(_.atZone(ZoneOffset.UTC).format(DateTimeFormatter.ISO_INSTANT))
+
+    client.expect[SearchListResponse](Request[F](Method.GET, Uri.unsafeFromString(baseUrl.toString() + "/search")
+      .withQueryParam("key", apiKey)
+      .withQueryParam("channelId", channelId)
+      .withQueryParam("q", searchString)
+      .withQueryParam("type", "video")
+      .withQueryParam("maxResults", 50)
+      .withOptionQueryParam("publishedAfter", formattedPublishAfter)
+      .withOptionQueryParam("pageToken", pageToken)))(jsonOf[F, SearchListResponse]).flatMap { response =>
+
+      fetchVideoDetails(response.items.map(_.id.videoId)).flatMap { videos =>
```

Comment:
Another nitpick, try to externalize the logic
`response.items.map(_.id.videoId)`
to a variable and provide the variable, it's more descriptive and easy to read